### ISSUE TITLE
Some notes on branch and tag names.

### ISSUE THREAD

**Issue by rye** [kotct/docket-old#41]
First of all, in the future, we should change how we do branches (to eliminate ambiguity). Ref-based clients (i.e. `magit`, which I know @cg505 and I use), choke when you have such a situation as having both a tag called `v0.2.0` and a branch called `v0.2.0`. I also think that all of the branch names should be more verbose.

For example, `fixes` should be renamed to `patch`, because all changes on `fixes` go into the next "patch" release. The branch `vx.x.x` should be renamed to `branch-vx.x.x`, and the tag syntax of `vx.x.x` should be changed to `tag-vx.x.x` or `release-vx.x.x`.

This prevents complaints or just general confusion.

As of now, however, I plan to do the following: I will delete branches, and, as we work, we will adopt this new strategy (unless, of course, a better one is suggested).

**Comment by Sammidysam** [comment]
I think it's a fine convention.

**Comment by Sammidysam** [comment]
Actually, @cg505 and I just talked about this and we prefer to keep the tag convention, but have different branch names.

**Comment by rye** [comment]
That makes sense.

**Comment by Sammidysam** [comment]
Maybe just name the branch for the next version as what the focus is?  That could work.

**Comment by rye** [comment]
Nah, I'd just change the different branch names. Maybe have a `next-major`, `next-minor`, and `next-patch` branch?

**Comment by Sammidysam** [comment]
I think that works.  Also, @cg505 just suggested that we could just have the branch names not have a prefix `v`.  What do you think of that?

**Comment by rye** [comment]
That doesn't make as much sense to me.

**Comment by Sammidysam** [comment]
Fine, then use the branches that you suggested (`next-major`, `next-minor`, `next-patch`).

**Comment by rye** [comment]
I actually like your idea better (hence my usage of it). What I'm doing is simply `branch-vx.x.x`, `branch-vx.x`, and (though it hasn't been created yet), `branch-vx`. All releases simply have the tag set as the suffix of the corresponding branch name (e.g. release for version 0.2.1 has tag `v0.2.1`).

**Comment by rye** [comment]
Hotfixes, or fixes for issues, that will get merged either directly into `master` or will go into their own `patch` release (either one, we don't have regulations to my knowledge) are named similarly: `hotfix-[description of issue separating words with hyphens]`.

**Comment by rye** [comment]
Hotfix branches should probably be deleted once they've been merged into master.

**Comment by Sammidysam** [comment]
Who are you speaking to here?  I also am not sure what was going on with communication.  I can't really even follow what I was saying.  Either way, I don't really like the branch convention, but whatever.  I can deal with it.  That was my plan earlier (just disliking but dealing with it).  My personal preference is still having the branch name correspond to what the focus of the branch is or what the branch is doing.  Otherwise, branch names tell pretty much nothing of what is actually being done on the branch.

**Comment by Sammidysam** [comment]
I think that the branch names should not correspond to the tag names.  The branches should make it seem like there are no versions, and then only the tags show that there are versions.

We have a bit too many branches right now, and I suggest far less.  A list of good branches to have is below.
- `master` (duh)
- `fixes`---minor fixes that, when merged, should be patch releases.  If we want to have a separate branch for cleaning up code, that would be `clean-up`, but this should suffice for all minor changes.
- A branch for the focus of the next release, named for the focus.  This could be named something like `user-configuration`, and branches could be made off of it whenever something needs its own branch.  We do _not_ need a branch for the next minor version.  A focus will suffice.

We need to delete branches when they are no longer in use.  It looks really bad to have 20 branches sitting around doing nothing.  We might accidentally obliterate code that is left there.

I recognize that this is the old convention, but I like it so much better.

**Comment by rye** [comment]
Okay. I agree; we should transition back to the old scheme.

**Comment by rye** [comment]
Since its contents have already been merged into master, I have deleted `branch-v0.2.2`.

**Comment by Sammidysam** [comment]
Hooray! :+1:

**Comment by rye** [comment]
Also, branch `v0.2.0` got confused with tag `v0.2.0` and so I had to euthanize both of them. I lost the contents of release `Version 0.2.0` in the process, but it worked out fine enough. I'll slowly be removing these branches in the future.

**Comment by Sammidysam** [comment]
That's bad.  I guess that means that we lost the description and may need to re-create it.

So, for the future, avoid at all costs having branch names correspond to releases?

**Comment by rye** [comment]
Absolutely. I'm not sure if you're familiar with the term "ref", but if there is more than one thing that uses the same name (illud est both a tag and a branch with the same name), git gets very confused and cannot do anything. As such, the only real solution to such a situation is to delete both and re-create as necessary. Preventative measures should be taken to prevent such a situation from happening, and using a version-less branch system is very handy for that.

**Comment by rye** [comment]
I don't think that the description of that release was that important, but for sentiment I suppose it's worth trying to recreate.

**Comment by rye** [comment]
I would suggest that we consider prematurely merge `branch-v0.2.3` and release Version 0.2.3 so that we can delete all of the unnecessary branches and adopt a version-branch-less system. We should keep organizational Issue Milestones for tracking what needs to be done in the next patch and use those for tracking our progress, if necessary. It seems that we should remove version-specific branches entirely since they overcomplicate things, and we should then just use feature branches, each of which gets merged (and, if complete, deleted) as they are added. Issue milestones should then be used to track the upcoming releases and the progress that has been made on them (I would say a maximum of four or five patch versions into the future should be the extent into the future for the patch milestones, one or two minors if necessary should the extent into the future for the minor milestones, and we should use the "Parking" milestone for ideas, miscellaneous, or deep-future ideas or issues).

In addition, I think some prefixes should be established for branch naming, and some ideas are listed below.
- **`hotfix-`** would be used for patches on critical bugs or features that severely cripple the project.
- **`fix-`** would be used for patches on bugs or features that cause problems but in rare use-cases or under certain conditions that do not immediately effect Production.

This system would separate the fixes from the features, which is the main idea of adopting this system. If we wanted to, we could create "feature pool" branches where the branch corresponds to a specific system (e.g. user configuration, authentication, etc.), and which would be used as the parent for any branches that work on that system. Such branches could have both `*fix-` and normal branches as children, so we'd be organizing by feature, not commit type.

In any case, we would need to establish a system for releasing versions. Ideally, issues corresponding to features or fixes would be closed as soon as the changes had been merged into master, and then as soon as a milestone is complete we would release. The tag would be created on the commit on `HEAD`, and then it would be ready for deploy.

I appreciate that we're experimenting with different version/branch systems on this project--it's good to get a good feel for what works best on projects like this. I have no problem with changing back and forth as long as we're learning from each switch and we have good reason for doing so.

**Comment by rye** [comment]
A good thing to consider would be adopting a version-less system (rolling-release). If we remove version branches and use feature branches, this makes sense. We'd just have a daily, weekly, biweekly, semiweekly, or monthly release schedule and we would just make a release on a pre-determined day and not worry about version numbering or such. There are certain advantages and disadvantages to this system, and I'm not suggesting this over the previous comment's suggestion, but I'm just naming it as a possibility.

**Comment by rye** [comment]
Such feature branches would either be kept or deleted, depending on completion and room to grow with the system.

**Comment by rye** [comment]
@cg505 / @Sammidysam, would releasing work on `branch-v0.2.3` to switch to the new system now be acceptable? Are there any problems with the aforementioned changes? Any suggestions?

**Comment by Sammidysam** [comment]
On the note of releasing 0.2.3 right now, no.  A better solution would be to simply rename your local branch to `fixes` (our miscellaneous fixes branch) via `git branch -m branch-v0.2.3 fixes` and then push that.  Once that is done, we can mess with the other branches without losing those three commits.  There is no point in releasing a new version when not any user-noticeable changes have been made.

Yes, the new branch system will be _feature-based_, not version-based.  We will make a branch for every major feature and put the miscellaneous minor items in `fixes`.  We will merge `master` to these feature branches whenever a new release comes out.  Branch naming is not a big deal, it just needs to provide some clue as to what is going on.  It needs to be short, though.  I feel like making a separate branch for each fix would become painful.  It would result in one branch for one commit often.  That is too much; that is why I have suggested a miscellaneous fixes branch `fixes`.  Imagine how many branches we would have as we forget to delete them while making that many branches.

Milestones aren't a big deal.  They are just convenient for keeping track of what we need to do, but there is no reason to make fancy rules for them.  Just make milestones for the minor version releases and assign issues accordingly.  We do not need milestones for patches because patches can just be released when we feel that the amount of fixes is high enough for a patch release or a very important fix was made.  "Parking" should not be a milestone because we should have our milestones be actual things that we can work for.  In that case, do not assign a milestone.  We should also only have versions as milestones; I do not see a point for anything else.  Milestones aren't used to categorize issues.

I completely agree with you with releasing new versions.  However, I think that we should not wait until the changes have been merged to master to close the issue for a problem.  I think that as soon as we agree that it is fixed on a branch, the corresponding issue can be closed.

To rolling releases, no.  That will result in too many unfinished or unstable features at release time.

With the branches, the goal should be to not have _too_ many.  We should not work on too many things at the same time.  I'd say 5--10 is okay, but much more will become a problem.

**Comment by rye** [comment]
The whole idea about my idea for the system is that branches **would not** be merged into `master` until ready, and they would only be deleted _once there is absolutely no progress to be made_. In the case of fixes, most people would just write one commit, however it is much better practice to make a branch for each fix. This indicates that you are working on a specific objective and it helps to guide you in the right direction, if you stray easily. Once a fix is merged into `master` and it's going to be included in the next release, the original creator of the branch (or someone who has permission from them) can delete the branch to keep things clean.

There are a number of advantages to having multiple `fix-` branches. As I said before, this indicates that all work on a specific branch corresponds to a specific objective. A branch like `fix-routing-error-for-docket` specifically states that all work on that branch corresponds to fixing the routing error for some "docket" thing--ideally, from within the commits the developer would reference the issue corresponding to the routing error, if such a thing had been reported. In addition, having branches like that gives the developer a simple idea of where things stand bug-wise. Since it seems like we're going with a `[branch]`-to-`master` merging scheme in that we're just merging directly into master to prepare for the next release, running `git branch` should (if everything is working properly) give the developer an idea of what kinds of fixes are currently in progress. In addition, if a "hotfix" gets fixed in a hacksy sort of way, either the branch can be renamed or a new "fix" branch can be created to replace it, using a more sturdy and sane approach.

Unfortunately, I strongly dislike the idea of using a `fixes` branch. Even single-commit changes can still warrant creating a new branch. It may be painful if the fix is only a one or two line change, but hopefully they'd be more than that and it'd be worth it. Creating a new branch has a number of added benefits, as well. For one, when you're creating the branch, you give yourself an idea of what it is that you want to do; by naming the branch something you are outlining what your exact objective is. Then, once you've created the branch, all of your work on that branch is separated from everyone else's work on their fixes, if there are multiple fixes in progress at once (unlikely, considering people's current contribution rates, but still). If you then merge that directly to master, there's no intermediate stage, which eliminates some problems.

My schema seems to be more common practice than having either a hub `fixes` branch or a general `fixes` branch for all fixes. For the first situation (utilizing a hub `fixes` branch), this is the case because having a hub `fixes` branch is unnecessary, and for the second situation (utilizing a general `fixes` branch for all fixes), this is the case because some work on some fixes may be at a different stage at a given time. In addition, merge conflicts can easily result from people working on the same file.

I agree about milestones, and I don't really care about milestones that much. We _do_ need to clean up our existing milestones and tags, and add a `parked` or `parking` tag to mark things as "parked" ideas.

Really, we can have as many branches as necessary. The point is that we not have ne quid nimis (nothing in excess) and that we be organized with our usage of branches. In regards to your comment that "we should not wait until the changes have been merged to master," I see your point. However, it seems that we all have somewhat different work schedules and so waiting for agreement could be time-consuming. I think if one is responsible and knows what the implications are of their merging a branch into `master`, they're good to go ahead and merge it. Even if it's a massive change, we still don't need to babysit each other. I think people should assign themselves to issues and then do their own work, then merge and mark as complete. If they need help, they can certainly ask for help, but I really think we don't need policy on agreement or any "we agree" sort of requirements for merges to take place. Hopefully people aren't writing controversial code, and hopefully they can use their good judgement to decide whether or not their "fix" that breaks the entire authentication system (for example) is good to roll. After all, that's what `git revert` is for :smile:.

**Comment by Sammidysam** [comment]
I realized something during the week while simulating how this branch system would work in my head.  If your proposed branch system is used, often times someone would fix something very quickly.  For example, they could create a local branch `fix-parenthesis-error` which contains a one character change, merge it into `master`, then push `master`.  It would look as if they pushed straight to `master`.  At this point, it seems as if just pushing stable fixes to `master` right away would be fine.  It seems fine to me.

However, before we can even get into the branch systems, we need to decide what the function of `master` is.  Is it a place to put the latest stable code, or a place to put the latest stable release?  If it is a place to put the latest stable code, then we need to merge to `master` much more often---when something is stable, merge away.  If it is a place to put the latest stable release, then we should keep merging items only when a release is ready.  I personally prefer the first option, as we can provide links to the latest release often, and somebody can get the latest stable code from GitHub.  It makes sense to me.  I am going to abstain from discussing the branch systems until we decide what the function of `master` is now.

I find milestones to be much more useful than tags.  Milestones help track progress towards the next release, whereas tags do just about nothing.  You can sort by them, but our tag system is already such a mess that sorting will not do any good.  Personally, the default tags typically suffice.

But yes, we should trust each other more.  I remember that I have read over commits a lot in the past, and we should really do that less (if others do the same).

**Comment by rye** [comment]
Your simulation is not necessarily accurate. Merging like that is called "fast-forwarding," and the commits are simply played back in a way onto the branch they're being merged to. If we do not fast-forward (i.e. we use `--no-ff` in *sh or `-nf` with magit), a commit is created. I would prefer if we did not use fast-forwards, which would make sense.

Again I say, hopefully people won't be merging unstable or crappy code with errors that can be fixed by one-character modifications into `master`. As @cg505 has suggested in the past, I'd like to mirror his suggestion of constructing a test suite and using it heavily. [Travis CI](https://travis-ci.org/) can run the tests off-site or we can run them on-site as we're editing, and the test system is capable of ensuring that our code works on the outside. We could then use pull requests (Travis CI latches on to those and checks to see if they'll work), however that seems to go against my suggestion.

I think `master`'s purpose should shift to being the latest stable code. We'd tag certain places on `master` as being releases, but everything would get merged to `master` as soon as it was stable. The end-user can download both the latest release from the releases system or the latest stable code. Again, all of the code on master would be the latest that works. @cg505 can then fetch the tags and checkout to the newest tag and deploy that (or we can just merge it to `heroku`).

**Comment by Sammidysam** [comment]
Good; I agree with the purpose of `master`.  We would always want the deployed version to be the latest release and not the latest stable code.  The latest stable code is stable, but the latest release is _stable_. :wink:

I guess avoiding fast-forwards could be done, though I am curious what would be so shameful about pushing directly to `master` for minor fixes.  We do it with [team461WBI/website](https://github.com/team461WBI/website), and it only causes hell when somebody works on (something that has a branch dedicated to it) on `master` instead of the branch that exists for that job (i.e. https://github.com/team461WBI/website/commit/10620cd7ccd6f399341f633ef013701033efb854, which should have gone on the `clean-up` branch, as it was a cleaning up change).  Besides that occurrence, it works completely fine.  For larger work, however, branches become needed.  I've seen pushing to `master` for minor work and branches for larger work used elsewhere.  It looks like it works fine to me.

Test suites sound fine.  I do not really want lots of pull requests, though.  We can, for the most part, test code ourselves (especially since we just run a command and magic happens).

**Comment by rye** [comment]
I would say that we should just try the `fix-[blah]` system. I agree that it doesn't make too much sense for small one-commit changes, however if people are committing often I find that sometimes even the smallest fixes end up spanning multiple commits. If you risk it and put code directly on `master`, you have a window where the code isn't fully functional. Besides, this also causes commits to get merged into `master` in chunks, which eliminates merge conflicts if two individuals are working on it at the same time. They resolve in chunks as opposed to repetitively. Therefore, I suggest that we work on committing more frequently (a good thing), having a specific goal in mind when we are committing (another good thing), using branches obsessively, and finally that we, in general, try to be more organized with our system.

**Comment by rye** [comment]
Since it appears that we are in agreement, after v0.2.3 is released, I plan to switch to this new branch system by either merging and deleting the old branches or renaming them to conform to the new schema. If, in the future, we encounter problems with this system, I'll be open to changing it, but I think this is probably one of the best ways we could go about organizing the system. We may decide to collect all of the small fixes into one branch, but I like the idea of keeping all of the fixes--no matter how small or how big--separate from each other (on separate branches). This prevents issues stemming from simultaneous work or incomplete work on modules and allows things to get merged in chunks of relevant commits. We can always change later, but I plan to switch to this new system once v0.2.3 is done.

**Comment by cg505** [comment]
I will also create a branch, `latest-release`, which will point to the same commit that the tag with the highest version number points to.

**Comment by cg505** [comment]
There is no reason to wait until v0.2.3 is released. I have merged branch-v0.2.3 into master, as our working (but stable) branch for the next patch release.

**Comment by rye** [comment]
Okay. I will then delete all of the inactive and no-longer-pertinent branches and start applying the new scheme.

**Comment by rye** [comment]
I'm glad to see that this project is active again.

**Comment by rye** [comment]
So this is what the project's branch schema looks like now:
- **`master`**: Latest stable code.
- **`fix-<error>`**: For fixing bugs and features that are mundane but still need to be fixed.
- **`hotfix-<error>`**: For fixing bugs and features that are very important and need to be fixed immediately.
- **`<feature>`**: Will house all work on **`<feature>`**.

**Comment by cg505** [comment]
Okay. I still don't see why (as I mentioned in #54) we need to avoid fast-forward merges. We don't need to know a merge happened. The point of having `fix-` and `hotfix-` branches is so that fixes and hotfixes that take more than one commit don't get mixed up in everything else. Even if we do feel it would be good practice to use them in cases of only one commit, there's no reason we need to be able to see that that happened.

**Comment by rye** [comment]
I'll tell you why. Fast-forward merges make it seem like the changes were committed directly to the branch that they're getting merged to. If people go back and look at it, then they don't see that a merge took place; they just see the changes. That's not okay. Having an actual merge commit helps to establish that a merge actually took place and makes it absolutely clear; fast-forward's don't. Also, since people love writing vague commit messages, that contributes to the problem. When I see a commit with the message "Merge branch 'foo' into 'bar'" then I know that a merge took place, but when I (and I know others would probably do the same) see a bunch of changes mixed in that don't have a place in the branch they're in and I _don't_ see a merge commit, I get confused. Merge commits help to alleviate that.

Of course, if no work has been done on a branch before you merge to it, non-fast-forwards don't help that much to prevent problems from happening, but if we're going to allow fast-forwards then people need to be less anal about the branch system and need to not complain if they think something doesn't conform to it.

**Comment by rye** [comment]
Also, having merge commits keeps the branches displayed on the network graph. That means that commits are kept separate by history, and that helps to cause distinguishment.

**Comment by cg505** [comment]
I see what you're saying. I don't really think its worth it, but I also don't really care, so whatever. If I screw it up and forget to disable fast-forwarding, sorry.
